sdks/go/pkg/beam/core/graph/coder/testutil/testutil.go

```diff
@@ -0,0 +1,136 @@
+	v.decBldSchema.Register(rt, decF)
+}
+
+// Validate is a test utility to validate custom schema coders generate
+// beam schema encoded bytes.
+//
+// Validate accepts the reflect.Type to register, factory functions for encoding and decoding, an
+// anonymous struct type equivalent to the encoded format produced and consumed by the factory produced functions
+// and test values. Values must be a single struct or pointer to struct.
+//
+// TODO(lostluck): Improve documentation.
+// TODO(lostluck): Abstract into a configurable struct, to handle
+//
+// Validate will register the under test factories and generate an encoder and decoder function.
+// These functions will be re-used for all test values. This emulates coders being re-used for all
+// elements within a bundle.
+//
+// Validate mutates the SchemaCoderValidator, so the SchemaCoderValidator may not be used more than once.
+func (v *SchemaCoder) Validate(t *testing.T, rt reflect.Type, encF, decF, schema interface{}, values interface{}) {
+	t.Helper()
+	testValues := reflect.ValueOf(values)
+	// Check whether we have a slice type or not.
+	if testValues.Type().Kind() != reflect.Slice {
+		vs := reflect.MakeSlice(reflect.SliceOf(testValues.Type()), 0, 1)
+		testValues = reflect.Append(vs, testValues)
+	}
+	if testValues.Len() == 0 {
+		t.Fatalf("No test values provided for ValidateSchemaCoder(%v)", rt)
+	}
+	// We now have non empty slice of test values!
+
+	v.encBldUT.Register(rt, encF)
+	v.decBldUT.Register(rt, decF)
+
+	testRt := testValues.Type().Elem()
+	encUT, err := v.encBldUT.Build(testRt)
+	if err != nil {
+		t.Fatalf("Unable to build encoder function with given factory: coder.RowEncoderBuilder.Build(%v) = %v, want nil error", rt, err)
+	}
+	decUT, err := v.decBldUT.Build(testRt)
+	if err != nil {
+		t.Fatalf("Unable to build decoder function with given factory: coder.RowDecoderBuilder.Build(%v) = %v, want nil error", rt, err)
+	}
+
+	schemaRt := reflect.TypeOf(schema)
+	encSchema, err := v.encBldSchema.Build(schemaRt)
+	if err != nil {
+		t.Fatalf("Unable to build encoder function for schema equivalent type: coder.RowEncoderBuilder.Build(%v) = %v, want nil error", rt, err)
+	}
+	decSchema, err := v.decBldSchema.Build(schemaRt)
+	if err != nil {
+		t.Fatalf("Unable to build decoder function for schema equivalent type: coder.RowDecoderBuilder.Build(%v) = %v, want nil error", rt, err)
+	}
+	for i := 0; i < testValues.Len(); i++ {
+		t.Run(fmt.Sprintf("%v[%d]", rt, i), func(t *testing.T) {
+			var buf bytes.Buffer
+			want := testValues.Index(i).Interface()
+			if err := encUT(want, &buf); err != nil {
+				t.Fatalf("error calling Under Test encoder[%v](%v) = %v", testRt, want, err)
+			}
+			initialBytes := clone(buf.Bytes())
+
+			bufSchema := bytes.NewBuffer(clone(initialBytes))
+
+			schemaV, err := decSchema(bufSchema)
+			if err != nil {
+				t.Fatalf("error calling Equivalent Schema decoder[%v]() = %v", schemaRt, err)
+			}
+			err = encSchema(schemaV, bufSchema)
```

Comment:
Does reading from the buffer (a few lines above with `decSchema`) remove those bytes from the buffer (or otherwise prevent those bytes from being cloned)? Because on first read it seems like `roundTripBytes` below will contain the element's bytes twice, the first copied from `initialBytes`, and the second from `encSchema` here.